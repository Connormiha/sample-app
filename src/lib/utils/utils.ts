export const promiseNoop = () => Promise.resolve();
export const noop = () => {};
export const delay = (time: number): Promise<void> =>
    new Promise((resolve) => {
        setTimeout(resolve, time);
    });

export interface IAsyncAction {
    cancel: () => void;
    promise: Promise<void>;
}

export interface IAsyncCallbacks<T> {
    success?: (res: T) => void;
    fail?: (err: any) => void;
    always?: () => void;
}

export const asyncAction = <T>(
    promise: Promise<T>,
    {success, fail, always}: IAsyncCallbacks<T>
): IAsyncAction => {
    let isCanceled = false;

    const newPromise = promise.then(
        (res) => {
            if (!isCanceled) {
                if (success) {
                    success(res);
                }
                if (always) {
                    always();
                }
            }
        },
        (err) => {
            if (!isCanceled) {
                if (fail) {
                    fail(err);
                }
                if (always) {
                    always();
                }
            }
        }
    );

    return {
        cancel: () => isCanceled = true,
        promise: newPromise
    };
};
