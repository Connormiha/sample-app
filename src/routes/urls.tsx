export const URL_SERVICES = '/services';
