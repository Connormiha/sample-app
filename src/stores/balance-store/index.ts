export * from './balance-store';
export {default} from './balance-store';
