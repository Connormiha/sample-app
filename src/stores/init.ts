import {configure} from 'mobx';
import NotificationsStore from 'stores/notifications-store';
import ServicesStore from 'stores/services-store';

configure({enforceActions: 'observed'});

export default () => ({
    notifications: new NotificationsStore(),
    services: new ServicesStore()
});
