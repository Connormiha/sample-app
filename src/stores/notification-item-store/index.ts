export * from './notification-item-store';
export {default} from './notification-item-store';
