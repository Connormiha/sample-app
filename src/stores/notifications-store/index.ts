export * from './notifications-store';
export {default} from './notifications-store';
