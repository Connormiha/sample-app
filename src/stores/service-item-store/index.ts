export * from './service-item-store';
export {default} from './service-item-store';
