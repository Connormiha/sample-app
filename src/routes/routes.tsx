import React from 'react';
import {BrowserRouter, Route, Redirect, Switch} from 'react-router-dom';
import PageServices from 'components/pages/page-services';
import PageNotFound from 'components/pages/page-not-found';
import * as urls from './urls';

export default (
    <BrowserRouter basename="/">
        <Switch>
            <Redirect from="/" exact to="/services" />
            <Route path={urls.URL_SERVICES} component={PageServices} />
            <Route path="*" component={PageNotFound} />
        </Switch>
    </BrowserRouter>
);
