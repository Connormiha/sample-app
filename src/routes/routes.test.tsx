import React from 'react';
import {Provider} from 'mobx-react';
import routes from 'routes/routes';
import stores from 'stores/init';
import {render, RenderResult} from 'react-testing-library';
import {delay} from 'lib/utils';

describe('Route', () => {
    const originalHref = location.href;
    const originalTitle = document.title;

    async function matchRoute(url: string, action?: (tree: RenderResult) => void): Promise<void> {
        history.replaceState({}, originalTitle, url);

        const tree = render(
            <Provider {...stores()}>
                {routes}
            </Provider>
        );

        await delay(0);

        if (action) {
            action(tree);
            await delay(0);
        }
        const fragment = tree.asFragment();
        tree.unmount();
        expect(fragment).toMatchSnapshot();
    }

    afterAll(() => {
        history.replaceState({}, originalTitle, originalHref);
    });

    type ICase = [string, string, ((tree: RenderResult) => void)?];
    const CASES: ICase[] = [
        [
            '/foo', '<PageNotFound />'
        ]
    ];

    CASES.forEach(([url, pageName, action]) => {
        it(`should render ${pageName}`, async () => {
            await matchRoute(url, action);
        });
    });
});
