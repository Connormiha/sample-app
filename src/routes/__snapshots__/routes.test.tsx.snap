// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route should render <PageNotFound /> 1`] = `
<DocumentFragment>
  <div
    class="layout-page-main"
  >
    <section
      class="layout-page-main__left"
    >
      <section
        class="layout-page-panel-menu"
      >
        <div
          class="layout-page-panel-menu__logo"
        >
          <span
            class="icon icon_name_logo icon_size_large"
          />
        </div>
        <div
          class="layout-page-panel-menu__menu"
        >
          <nav>
            <a
              class="menu-item"
              href="/some-servive-1"
              title="Сервис 1"
            />
            <a
              class="menu-item"
              href="/some-servive-2"
              title="Сервис 2"
            />
            <a
              class="menu-item"
              href="/some-servive-3"
              title="Сервис 3"
            />
            <a
              class="menu-item"
              href="/some-servive-4"
              title="Сервис 4"
            />
            <a
              class="menu-item"
              href="/services"
              title="Сервисы"
            />
            <a
              class="menu-item"
              href="/some-servive-5"
              title="Сервис 5"
            />
            <a
              class="menu-item"
              href="/some-servive-6"
              title="Сервис 6"
            />
          </nav>
        </div>
      </section>
    </section>
    <section
      class="layout-page-main__right"
    >
      <header>
        <div
          class="layout-page-header"
        >
          <div />
          <div
            class="layout-page-header__right"
          >
            .
          </div>
        </div>
      </header>
      <main
        class="layout-page-main__content"
      >
        Несуществующая страница
      </main>
    </section>
  </div>
</DocumentFragment>
`;
